File: type-strike-android/app/src/main/java/com/typestrike/data/model/Leaderboard.kt
package com.typestrike.data.model

import com.google.gson.annotations.SerializedName

/**
 * A single entry on the leaderboard — matches backend LeaderboardEntry.
 */
data class LeaderboardEntry(
    @SerializedName("player_id") val playerId: Int,
    @SerializedName("player_name") val playerName: String,
    val level: Int,
    val xp: Int,
    @SerializedName("total_stars") val totalStars: Int,
    @SerializedName("levels_cleared") val levelsCleared: Int,
    @SerializedName("best_wpm") val bestWpm: Int,
    @SerializedName("updated_at") val updatedAt: String,
    val rank: Int
)

/**
 * Request body to sync a player's stats to the leaderboard.
 */
data class SyncLeaderboardRequest(
    @SerializedName("player_id") val playerId: Int
)

/**
 * Response wrapping a list of leaderboard entries — matches backend LeaderboardResponse.
 */
data class LeaderboardResponse(
    val entries: List<LeaderboardEntry>,
    @SerializedName("total_count") val totalCount: Int,
    @SerializedName("player_rank") val playerRank: PlayerRankResponse? = null
)

/**
 * A specific player's rank with nearby competitors — matches backend PlayerRank.
 */
data class PlayerRankResponse(
    val entry: LeaderboardEntry,
    val above: List<LeaderboardEntry>,
    val below: List<LeaderboardEntry>
)

File: backend/internal/models/leaderboard.go
package models

import "time"

// LeaderboardEntry represents a player's current standing on the leaderboard.
// This is a query result from leaderboard_entries view/table — no GORM model tag needed.
type LeaderboardEntry struct {
	PlayerID      int       `json:"player_id"`
	PlayerName    string    `json:"player_name"`
	Level         int       `json:"level"`
	XP            int       `json:"xp"`
	TotalStars    int       `json:"total_stars"`
	LevelsCleared int       `json:"levels_cleared"`
	BestWPM       int       `json:"best_wpm"`
	UpdatedAt     time.Time `json:"updated_at"`
	Rank          int       `json:"rank"`
}

// LeaderboardResponse wraps a list of leaderboard entries with metadata.
type LeaderboardResponse struct {
	Entries    []LeaderboardEntry `json:"entries"`
	TotalCount int                `json:"total_count"`
	PlayerRank *PlayerRank        `json:"player_rank,omitempty"`
}

// PlayerRank shows a specific player's rank with nearby competitors.
type PlayerRank struct {
	Entry LeaderboardEntry   `json:"entry"`
	Above []LeaderboardEntry `json:"above"`
	Below []LeaderboardEntry `json:"below"`
}
